### PR TITLE
Deprecate |>

### PR DESCRIPTION
**Pull request by bramtayl**

### CODE REVIEW

**Comment by bramtayl** [comment]
Not sure if this is fully sufficient. I think might need to be deprecated to a syntax error if we want non-function-like functionality in the future?

**Comment by shashi** [comment]
This has been discussed before, and the consensus was against removing `|>`. https://github.com/JuliaLang/julia/issues/20331

**Comment by bramtayl** [comment]
Recent time of support here: https://github.com/JuliaLang/julia/issues/5571#issuecomment-346415577

**Comment by StefanKarpinski** [comment]
The purpose of deprecating it as a function is to reintroduce it as syntax.

**Comment by shashi** [comment]
Okay, sounds good.

**Comment by yurivish** [comment]
Would reintroducing it as syntax mean that you wouldn't be able to alias it, e.g. say `| = |>` in order to use a different spelling (say at the REPL)?

**Comment by bramtayl** [comment]
Not sure if the broadcast test i deleted was important...

**Comment by bramtayl** [comment]
Worth noting how much uglier code got without chaining (and how much nicer `|> x -> ` expressions could be with syntactic chaining.

**Comment by KristofferC** [comment]
To me, the code became quite a lot clearer.

**Comment by Sacha0** [comment]
Echoing Kristoffer, almost every rewrite in this pull request was clearer to me. Best!

**Comment by bramtayl** [comment]
Does anyone know what these deserializing a remote exception errors mean? I probably made a mistake in the triangular and cholmod files but I don't see a clear line number

**Comment by bramtayl** [comment]
da da da da da da daaaaaa

**Comment by bramtayl** [comment]
This looks good to me? CI seems unusually happy

**Comment by stevengj** [comment]
I have to say, I don't really understand the reason for deprecating this in order to "turn it into syntax".  The proposed syntax in #5571 seems to be mostly about currying, not piping.     Currying is useful in lots of cases besides piping, so I don't think currying syntax should be tied to a `|>` operator at all.

**Comment by bramtayl** [comment]
Right now `_` is also deprecated, but if I understand https://github.com/JuliaLang/julia/issues/5571#issuecomment-345991770 correctly, it would given meaning only in combination with `|>` so that it is short-hand for the left side. So

```
1 |>
   _ + 1 |>
   _ + 1
```

Would be 3.

**Comment by bramtayl** [comment]
Here's a particularly horrible line from this PR:

````
A1 = t1(elty1 == Int ? rand(1:7, n, n) : convert(Matrix{elty1}, (t -> uplo1 == :U ? t : adjoint(t))((t -> chol(t't))(elty1 <: Complex ? complex.(randn(n, n), randn(n, n)) : randn(n, n)))))
```

Here's how it might be nicer chained:

```
A1 = 
    randn(n, n) |>
    elty1 <: Complex ? complex.(_, _) : _ |>
    chol(_' * _) |>
    uplo1 == :U ? _ : adjoint(_) |>
    convert(Matrix{elty1}, _) |>
    t1(elty1 == Int ? rand(1:7, n, n), _)
```

**Comment by bramtayl** [comment]
Unless I'm mistaken, this would be might trickier to do with tight currying alone.

**Comment by KristofferC** [comment]
That chained stuff is completely unreadable to me.

**Comment by bramtayl** [comment]
I guess as someone who's been doing chaining in R for years the unchained alternative is completely unreadable.

**Comment by Sacha0** [comment]
Both versions are difficult to grok. Better to write such expressions out more clearly if more verbosely. Best!

**Comment by stevengj** [comment]
Giving `_` a meaning only in association with `|>` seems like a huge mistake to me.   A terse partial-application syntax for small functions like `f(_,y)` or `_.a` is useful in tons of cases that don't involve chaining.

A terse anonymous-function syntax for more complicated case, just to save writing `x ->...`, seems less important.

**Comment by bramtayl** [comment]
I guess I'd be ok with tight currying too. Vanilla |> and chaining would look like this:

```
A1 = 
    randn(n, n) |>
    _ -> elty1 <: Complex ? complex.(_, _) : _ |>
    _ -> _' * _ |>
    chol |>
    _ -> uplo1 == :U ? _ : adjoint(_) |>
    convert(Matrix{elty1}, _) |>
    _ -> elty1 == Int ? rand(1:7, n, n) : _ |>
    t1
```

This is a bit of a mess. I'm much more of fan of

```
randn(n, n)
elty1 <: Complex ? complex.(it, it) : it
it' * it
chol(it)
uplo1 == :U ? it : adjoint(it)
convert(Matrix{elty1}, it)
elty1 == Int ? rand(1:7, n, n) : it
t1(it)
A1 = it
```

ala

https://github.com/JuliaLang/julia/issues/24082

**Comment by stevengj** [comment]
I agree with @Sacha0 that this example is hideous as a single expression of any kind; it seems better to rewrite it entirely, breaking it into multiple lines or even multiple functions.

**Comment by bramtayl** [comment]
I'm with you on that. Long chains are definitely confusing and breaking them into shorter chains is usually a good idea.

**Comment by stevengj** [comment]
@bramtayl, it's not so much the length of the chain in this case as all the conditionals.   If these conditionals were split into appropriate functions (and probably handled by dispatch, not by runtime checks), it would be a lot more comprehensible:
```jl
if elty1 == Int
   A = rand(1:7, n, n)
else
   A = randn(n,n) |> complexify(elty1, _) |> (x -> x'*x) |> chol |> fliptri(uplo1, _) |> Matrix{elty1}
end
t1(A)
```

**Comment by bramtayl** [comment]
Yep that looks nice too. No argument from me about splitting out commonly used patterns into functions or that if/else sometimes looks a lot nicer than ?. Even with the simplifications you made, `(x -> x'*x)` would be nicer as `_' * _` unless you wanted to split that out into a function too.

**Comment by StefanKarpinski** [comment]
When multiple `_` are used such as `_'_` it is much more common in languages that have these mini closure features to interpret it as `(x,y) -> x'y` rather than as `x -> x'x`.

File: doc/src/manual/mathematical-operations.md [hunk 1]
```diff
@@ -361,7 +361,6 @@ Julia applies the following order and associativity of operations, from highest
 | Bitshifts      | `<< >> >>>`                                                                                       | Left                       |
 | Addition       | `+ - \| ⊻`                                                                                        | Left[^2]                    |
 | Syntax         | `: ..`                                                                                            | Left                       |
-| Syntax         | `\|>`                                                                                             | Left                       |
 | Syntax         | `<\|`                                                                                             | Right                      |
```
Comment (kmsquire) [inline doc/src/manual/mathematical-operations.md]: If `|>` is being removed, shouldn't `<|` also be removed here?

Comment (bramtayl) [reply doc/src/manual/mathematical-operations.md]: Hmm. Not sure. Why?

**Comment by bramtayl** [comment]
I guess I've been convinced by @stevengj that `_` is more important for use in tight lambdas than for chaining. I think that `|>` combined with tight lambdas is nice but really only covers a very small subset of what you can use chaining for. If we want chaining in proper Base we will have to use a different set of symbol. All that we need for proper chaining is a stand-in for the previous evaluation (something other than `_`) deprecated for 0.1. I'm curious about @oxinabox's thoughts.

File: doc/src/manual/mathematical-operations.md (hunk 1 above)
Comment (bramtayl) [reply doc/src/manual/mathematical-operations.md]: Oh I see. Yeah dunno.

**Comment by stevengj** [comment]
In cases where a “tight” `_` doesn’t work, I’m not convinced that `x ->` is so onerous.

**Comment by bramtayl** [comment]
If that's the case, then why `_` as a lambda at all?

**Comment by oxinabox** [comment]
>I guess I've been convinced by @stevengj that _ is more important for use in tight lambdas than for chaining. 

I 100% agree with that. that having terse lambdas is what the `_` proposal is about.
It is an end-in-and-of-itself.
Separately from chaining.
The fact that it would make chaining nicer is a side effect.
Its use in map, reduce, and other higher order functions is what is really attractive.

I'm not a huge fan of chaining these days. (It has its place sure)


> I think that |> combined with tight lambdas is nice but really only covers a very small subset of what you can use chaining for. 

I'm not sure what in particular you are thinking of.

> If we want proper chaining in Base we will have to use a different symbol. 

Ergo this deprecation, right?

**Comment by stevengj** [comment]
I don't support this deprecation, because I think a `_` syntax for short (single-call) anonymous functions combined with `x->` for longer anonymous functions is sufficient support for chaining in Julia.  `|>` seems like it should continue to be an ordinary operator.

I don't like the idea of a special currying syntax tied to `|>`, [as I've explained elsewhere](https://github.com/JuliaLang/julia/pull/24990#issuecomment-350508958), and supporting such a syntax in the future seems to be the whole reason for this deprecation.

**Comment by bramtayl** [comment]
`|>` isn't necessary for chaining, so I'd be fine with giving it up to currying.

**Comment by StefanKarpinski** [comment]
Triage: instead of deprecating, we'll document that `|>` may become syntax in the future such that using it for pipelining will continue to work as it currently does but overloading it may break.

**Comment by oxinabox** [comment]
It might be nice to cherry-pick this PR, and keep a lot of the edits to remove `|>`.
I would say at least half of the code in base cleaner without it.

**Pull request closed without merging**